src/relay/pass/type_solver.cc

```diff
@@ -38,6 +39,131 @@ class TypeSolver::Reporter : public TypeReporterNode {
   TypeSolver* solver_;
 };
 
+class TypeSolver::Unifier : public TypeFunctor<Type(const Type&, const Type&)> {
+ public:
+  explicit Unifier(TypeSolver* solver) : solver_(solver) {}
+
+  Type Unify(const Type& src, const Type& dst) {
+    // Known limitation
+    // - handle shape pattern matching
+    TypeNode* lhs = solver_->GetTypeNode(dst);
+    TypeNode* rhs = solver_->GetTypeNode(src);
+
+    // do occur check so we don't create self-referencing structure
+    if (lhs->FindRoot() == rhs->FindRoot()) {
+      return lhs->resolved_type;
+    }
+    if (lhs->resolved_type.as<IncompleteTypeNode>()) {
+      solver_->MergeFromTo(lhs, rhs);
+      return rhs->resolved_type;
+    } else if (rhs->resolved_type.as<IncompleteTypeNode>()) {
+      solver_->MergeFromTo(rhs, lhs);
+      return lhs->resolved_type;
+    } else {
+      Type resolved =
+        this->VisitType(lhs->resolved_type, rhs->resolved_type);
+      CHECK(resolved.defined())
+        << "Unable to unify parent types: "
+        << lhs->resolved_type << " and " << rhs->resolved_type;
+      return resolved;
+    }
+  }
+
+  // child type needs to be listed in parent's relations, even though
+  // the child is not an argument to the relations (still have to
+  // update the relations if the child changes)
+  void RegisterChildType(const Type& parent, const Type& child) {
+    TypeNode* parent_node = solver_->GetTypeNode(parent);
+    TypeNode* child_node = solver_->GetTypeNode(child);
+
```

Comment:
I actually think this is incorrect behavior, because we will want to transfer relation lists more than one level down (e.g., in a case like `Unify(Tuple(Tuple(?a, ?b), ?c), Tuple(Tuple(?d, ?e), ?f)))`). I will investigate why there was hanging before and see if I can fix that (or get started on tests).

Reply:
(Figured out the problem, I was creating a circular reference because I was not allocating copies of list nodes; I will push a fix shortly.)